### PR TITLE
FO-2135 New identity python library

### PR DESCRIPTION
**Pull request by StephanieBracaloni**

### CODE REVIEW

File: iotics/lib/identity/register/document_helper.py [hunk 1]
```diff
@@ -15,6 +15,11 @@
 class RegisterDocumentHelper:
     @staticmethod
     def get_owner_public_key(doc: RegisterDocument) -> Optional[RegisterPublicKey]:
+        """
+        Get the public key that matches the DID identifier
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]: `Get the register document initial owner public key` ?

File: iotics/lib/identity/register/document_helper.py [hunk 2]
```diff
@@ -23,18 +28,41 @@ def get_owner_public_key(doc: RegisterDocument) -> Optional[RegisterPublicKey]:
...
+        :param doc: existing register document
+        :param include_auth: include authentication keys
+        :return: True/False if found
+        """
         return RegisterDocumentHelper.get_issuer_key(issuer_name, doc, include_auth) is not None
 
     @staticmethod
     def get_issuer_key(issuer_name: str, doc: RegisterDocument, include_auth: bool) -> Optional[RegisterKey]:
+        """
+        Find an issuer key
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]:
```
Find a register key by issuer.
Lookup in the register document public keys (and authentication keys if include_auth is set to True).
```

File: iotics/lib/identity/register/document_helper.py [hunk 3]
```diff
@@ -23,18 +28,41 @@ def get_owner_public_key(doc: RegisterDocument) -> Optional[RegisterPublicKey]:
...
             if key.base58 == public_base58:
-                return Issuer.build(issuer_did, key.name)
+                return Issuer.build(doc.did, key.name)
         return None
 
     @staticmethod
-    def is_issuer_in_keys(issuer_name: str, doc: RegisterDocument, include_auth: bool):
+    def is_issuer_in_keys(issuer_name: str, doc: RegisterDocument, include_auth: bool) -> bool:
+        """
+        Find an issuer key and returns True/False if found
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py:50]:
```
Find a register key by issuer and returns True/False if found.
Lookup in the register document public keys (and authentication keys if include_auth is set to True).
```

File: iotics/lib/identity/register/document_helper.py [hunk 4]
```diff
@@ -43,7 +71,15 @@ def get_issuer_key(issuer_name: str, doc: RegisterDocument, include_auth: bool)
         return None
 
     @staticmethod
-    def get_issuer_key_from_delegation(issuer_name: str, doc: RegisterDocument, include_auth: bool):
+    def get_issuer_key_from_delegation(issuer_name: str, doc: RegisterDocument,
+                                       include_auth: bool) -> Optional[RegisterDelegationProof]:
+        """
+        Find an issuer key for delegation
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]:
```
Find a register delegation proof by issuer.
Lookup in the register document control delegation proofs (and authentication delegation proofs if include_auth is set to True).
```

File: iotics/lib/identity/register/document_helper.py [hunk 5]
```diff
@@ -54,6 +90,13 @@ def get_issuer_key_from_delegation(issuer_name: str, doc: RegisterDocument, incl
     @staticmethod
     def get_delegation_by_controller(issuer: Issuer, doc: RegisterDocument,
                                      include_auth: bool) -> Optional[RegisterDelegationProof]:
+        """
+        Find delegation given controller issuer
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]:
```
Find a register delegation proof by controller issuer.
Lookup in the register document control delegation proofs (and authentication delegation proofs if include_auth is set to True).
```

File: iotics/lib/identity/register/document_helper.py (hunk 2 above)
Comment (StephanieBracaloni) [reply iotics/lib/identity/register/document_helper.py:60]: this function could be renamed `get_delegation_proof_by_issuer`

File: iotics/lib/identity/register/document_helper.py [hunk 6]
```diff
@@ -65,6 +108,17 @@ def get_delegation_by_controller(issuer: Issuer, doc: RegisterDocument,
     @staticmethod
     def get_valid_issuer_key_for(doc: RegisterDocument, issuer_name: str, get_controller_doc: GetControllerDocFunc,
                                  include_auth: bool, ) -> Optional[IssuerKey]:
+        """
+        Get an IssuerKey matching issuer name
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]:
```
Get a valid issuer key matching issuer name
```

File: iotics/lib/identity/register/document_helper.py [hunk 7]
```diff
@@ -80,9 +134,29 @@ def get_valid_issuer_key_for(doc: RegisterDocument, issuer_name: str, get_contro
     @staticmethod
     def get_valid_issuer_key_for_control_only(doc: RegisterDocument, issuer_name: str,
                                               get_controller_doc: GetControllerDocFunc) -> Optional[IssuerKey]:
+        """
+        Get an IssuerKey matching issuer name (control only)
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]: `Get a valid issuer key matching issuer name from the control keys and delegation proofs only`

File: iotics/lib/identity/register/document_helper.py [hunk 8]
```diff
@@ -80,9 +134,29 @@ def get_valid_issuer_key_for(doc: RegisterDocument, issuer_name: str, get_contro
...
+        :raises:
+            IdentityValidationError: if invalid name or did
+        """
         return RegisterDocumentHelper.get_valid_issuer_key_for(doc, issuer_name, get_controller_doc, include_auth=False)
 
     @staticmethod
     def get_valid_issuer_key_for_auth(doc: RegisterDocument, issuer_name: str,
                                       get_controller_doc: GetControllerDocFunc) -> Optional[IssuerKey]:
+        """
+        Get an IssuerKey matching issuer name (control and authentication)
```
Comment (StephanieBracaloni) [inline iotics/lib/identity/register/document_helper.py]: `Get a valid issuer key matching issuer name from the control and authentication keys and delegation proofs`

**Review by timchurchard — APPROVED** [review]

File: iotics/lib/identity/api/advanced_api.py [hunk 9]
```diff
@@ -96,120 +178,369 @@ def _update_doc(self, doc_owner_key_pair: KeyPair, doc_owner_issuer: Issuer, get
...
+            IdentityResolverError: if invalid resolver response
+            IdentityResolverDocNotFoundError: if document not found
+            IdentityResolverTimeoutError: if timeout error
+            IdentityResolverCommunicationError: if communication error
+        """
         return self.resolver_client.get_document(doc_did)
 
-    def get_document_if_exists(self, did: str) -> Optional[RegisterDocument]:
+    def get_document_if_exists(self, doc_did: str) -> Optional[RegisterDocument]:
+
```
Comment (timchurchard) [inline iotics/lib/identity/api/advanced_api.py]: blank line

File: iotics/lib/identity/api/advanced_api.py [hunk 10]
```diff
@@ -96,120 +178,369 @@ def _update_doc(self, doc_owner_key_pair: KeyPair, doc_owner_issuer: Issuer, get
...
+            IdentityValidationError: if invalid secrets
+            IdentityValidationError: if invalid name
+            IdentityInvalidDocumentError: if document build error
+            IdentityResolverConflictError: register document already exists with different owners
+            IdentityResolverError: if can not interact with the resolver
+            IdentityDependencyError: if incompatible library dependency
+        """
         return self.new_registered_identity(DIDType.TWIN, key_pair_secrets, name, override_doc)
 
     def verify_register_document(self, doc: RegisterDocument):
```
Comment (timchurchard) [inline iotics/lib/identity/api/advanced_api.py]: misnamed function verify/validate ?

File: iotics/lib/identity/api/advanced_api.py [hunk 11]
```diff
@@ -238,48 +610,134 @@ def add_control_delegation_proof_to_document(self, proof: Proof,
                                                  delegation_name: str,
                                                  doc_owner_issuer: Issuer,
                                                  doc_owner_key_pair: KeyPair):
+        """
+       Add register control delegation proof to a register document.
```
Comment (timchurchard) [inline iotics/lib/identity/api/advanced_api.py]: indent wrong

File: iotics/lib/identity/validation/authentication.py [hunk 12]
```diff
@@ -14,6 +14,26 @@ class IdentityAuthValidation:
     @staticmethod
     def is_allowed_for(issuer: Issuer, issuer_doc: RegisterDocument, subject_doc: RegisterDocument,
                        include_auth: bool):
```
Comment (timchurchard) [inline iotics/lib/identity/validation/authentication.py]: no return type?

**Review by timchurchard — APPROVED** [review]

File: iotics/lib/identity/api/high_level_api.py [hunk 13]
```diff
@@ -85,16 +134,42 @@ def create_twin_with_control_delegation(self, twin_seed: bytes, twin_key_name: s
     def get_ownership_of_twin_from_registered_identity(self, twin_registered_identity: RegisteredIdentity,
                                                        new_owner_registered_identity: RegisteredIdentity,
                                                        new_owner_key_name: str):
+        """
+        Get Ownership of a twin.
```
Comment (timchurchard) [inline iotics/lib/identity/api/high_level_api.py]: perhaps work saying like "get ownership of a twin by adding a key you control" or something

File: iotics/lib/identity/api/high_level_api.py [hunk 14]
```diff
@@ -85,16 +134,42 @@ def create_twin_with_control_delegation(self, twin_seed: bytes, twin_key_name: s
...
+            IdentityInvalidDocumentError: if invalid register document
+            IdentityResolverError: if resolver error
+        """
         new_owner_key_pair = KeyPairSecretsHelper.get_key_pair(new_owner_registered_identity.key_pair_secrets)
         current_owner_key_pair = KeyPairSecretsHelper.get_key_pair(twin_registered_identity.key_pair_secrets)
         self.advanced_api.add_public_key_to_document(new_owner_key_name, new_owner_key_pair.public_base58,
                                                      current_owner_key_pair, twin_registered_identity.issuer)
 
     @staticmethod
     def create_seed(length: Optional[int] = 128) -> bytes:
```
Comment (timchurchard) [inline iotics/lib/identity/api/high_level_api.py]: default can be 256?

File: iotics/lib/identity/api/regular_api.py [hunk 15]
```diff
@@ -55,24 +90,78 @@ def _get_identity(key_pair_secrets: KeyPairSecrets, did: str, name: str) -> Regi
...
+        """
         key_pair_secrets = build_twin_secrets(key_seed, key_name, seed_method, password)
         return self._get_identity(key_pair_secrets, twin_did, twin_name)
 
     def user_delegates_authentication_to_agent(self, user_registered_identity: RegisteredIdentity,
                                                agent_registered_identity: RegisteredIdentity,
                                                delegation_name: str):
+        """
+        User delegates authentication to agent.
+        The agent cant authenticate on behalf of the user.
```
Comment (timchurchard) [inline iotics/lib/identity/api/regular_api.py]: cant?

**Review by timchurchard — APPROVED** [review]

**Pull request merged**